File: PythonDatabaseCompatibility/varMethods.py
import ibm_db
import json
import inspect
import os
import sys
import functools

def immediateOper(connection, sql, outFile):
    outFile.write("Performing immediate operation: {} \n".format(sql))
    stmt = None
    try:
        stmt = ibm_db.exec_immediate(connection, sql)
    except:
        outFile.write(ibm_db.stmt_error())
        outFile.write(ibm_db.stmt_errormsg())
        raise
    if not stmt:
        outFile.write("Operation FAILED \n")
    return stmt

def setAutoComOff(connection, outFile):
    ibm_db.autocommit(connection, ibm_db.SQL_AUTOCOMMIT_OFF)
    # will return current state of autocommit, should be 0
    ac = ibm_db.autocommit(connection)
    if ac == 0:
        outFile.write("Autocommit set to off \n")
    if ac != 0:
        outFile.write("Cannot set autocommit to off. FAIL \n")

# verify that database is logging
def rollbackTrans(connection, outFile):
    if ibm_db.rollback(connection):
        outFile.write("Rollback successful \n")
    else:
        outFile.write("Rollback FAILED \n")

def commitTrans(connection, outFile):
    if ibm_db.commit(connection):
        outFile.write("Commit successful \n")
    else:
        outFile.write("Commit FAILED \n")

def performPrepStmtBindings(connection, outFile, jLine, preparedStmtList):
    currentBindings = jLine.bindingValueTuple
    outFile.write("Values to bind to current prepared statement: {} \n".format(currentBindings))
    currentSqlOper = preparedStmtList.getStmtIdSql(jLine.statementId)
    outFile.write("Executing sql statement: {} \n".format(currentSqlOper))
    currentHdl = preparedStmtList.getCurrentHdl(jLine.statementId)
    stmt = ibm_db.execute(currentHdl, currentBindings)
    if not stmt:
        outFile.write("operation FAILED")    

def performPreparedStmt(connection, outFile, jLine, preparedStmtList):
    currentSqlOper = getPreparedStmtSql(jLine.statementId, preparedStmtList)
    outFile.write("Executing sql statement: {} \n".format(currentSqlOper))    
    currentHdl = getPreparedStmtHdl(jLine.statementId, preparedStmtList)
    if not ibm_db.execute(currentHdl):
        outFile.write("operation FAILED")    

def prepareBindings(connection, outFile, jLine, preparedStmtList):
    outFile.write("Binding values: {} \n".format(jLine.bindings))
    currentSqlOper = getPreparedStmtSql(jLine.statementId, preparedStmtList)
    currentHdl = getPreparedStmtHdl(jLine.statementId, preparedStmtList)                 
    for paramDict in jLine.bindings:
        stmt = ibm_db.bind_param(currentHdl, paramDict['index'], paramDict['value'])
        if not stmt:
            outFile.write("operation FAILED")

def expectedEqualsActual(currentHdl, jLine, isOrdered, outFile):
    expectedResultsList = jLine.expectedResults
    fetchNum = jLine.nfetch
    actualResultList = getActualResultList(currentHdl, fetchNum)
    outFile.write("Expected Results: {} \n".format(expectedResultsList))
    outFile.write("Actual Results: {} \n".format(actualResultList))    
    if (actualResultList == [] and expectedResultsList == []):
        return True
    if isOrdered:
        return actualResultList == expectedResultsList
    else:
        found = False 
        for entry in actualResultList:
            for dictRow in expectedResultsList:
                if dictRow == entry:
                    expectedResultsList.remove(dictRow)
                    found = True
                    break  
            if not found:
                break
        if len(expectedResultsList) > 0:
            found = False
        if found:
            return True
        else:
            return False        
    
def getActualResultList(stmtHdl, fetchNum):
    resultList = []
    if fetchNum == None:
        resultDict = ibm_db.fetch_assoc(stmtHdl)
        if not resultDict:
            return resultList
        while resultDict:
            resultList.append(resultDict)
            resultDict = ibm_db.fetch_assoc(stmtHdl)
        return resultList
    else:
        for n in range(fetchNum):
            resultDict = ibm_db.fetch_assoc(stmtHdl)
            resultList.append(resultDict)
        return resultList
   
def getPreparedStmtSql(jLineStmtID, preparedStmtList):
    for preparedStmt in preparedStmtList:
        if preparedStmt.stmtId == jLineStmtID:
            return preparedStmt.sqlStmt


def getPreparedStmtHdl(jLineStmtID, preparedStmtList):
    for preparedStmt in preparedStmtList:
        if preparedStmt.stmtId == jLineStmtID:
            return preparedStmt.stmtHdl

def getPreparedStmtObj(jLineStmtID, preparedStmtList):
    for preparedStmt in preparedStmtList: 
        if preparedStmt.stmtId == jLineStmtID:
            return preparedStmt

def removePreparedStmtFromList(jLineStmtID, preparedStmtList):
    for preparedStmt in preparedStmtList:
        if preparedStmt.stmtId == jLineStmtID:
            ibm_db.free_stmt(preparedStmt.stmtHdl)
            preparedStmtList.remove(preparedStmt)
            
def enterResultToJsonFile(fileName, result, outfile, comment = None):
    jsonDict = {}
    jsonDict['language'] = 'python'
    jsonDict['client'] = 'IBM DB2 ODBC DRIVER'
    jsonDict['server'] = 'Informix 12.10'
    jsonDict['test'] = fileName
    jsonDict['result'] = result
    jsonDict['detail'] = comment        
    json.dump(jsonDict, outfile)
    outfile.write("\n")
    
def enterExtraDataToJsonFile(outfile):
    jsonDict = {}
    jsonDict['language'] = 'python'
    jsonDict['client'] = 'IBM DB2 ODBC DRIVER'
    jsonDict['server'] = 'Informix 12.10'
    jsonDict['test'] = "dataTypeTest_INT8.json"
    jsonDict['result'] = False
    jsonDict['detail'] = "int8 not compatible"        
    json.dump(jsonDict, outfile)
    outfile.write("\n")
    jsonDict = {}
    jsonDict['language'] = 'python'
    jsonDict['client'] = 'IBM DB2 ODBC DRIVER'
    jsonDict['server'] = 'Informix 12.10'
    jsonDict['test'] = "dataTypeTest_SERIAL8.json"
    jsonDict['result'] = False
    jsonDict['detail'] = "serial8 not compatible"       
    json.dump(jsonDict, outfile)
    outfile.write("\n")
    jsonDict = {}
    jsonDict['language'] = 'python'
    jsonDict['client'] = 'IBM DB2 ODBC DRIVER'
    jsonDict['server'] = 'Informix 12.10'
    jsonDict['test'] = "ops.json"
    jsonDict['result'] = True
    jsonDict['detail'] = "format of test is different than others, but operations performed in other tests"       
    json.dump(jsonDict, outfile)
    outfile.write("\n")
    
def getMyPath():
    try:
        fileName = __file__
    except NameError: # fallback
        fileName = inspect.getsourcefile(getMyPath)
    return os.path.realpath(fileName)

def getJsonTestDirectory():
    myPath = getMyPath()
    upperDirPath = functools.reduce(lambda x, f: f(x), [os.path.dirname]*2, myPath)
    testParentPath = os.path.join(upperDirPath, "JavaDatabaseCompatibility")
    return os.path.join(testParentPath, "resources")

File: RubyDatabaseCompatibility/varMethods.rb
module VarMethods

	module_function

	def immediateOper(connection, sql, outFile)
		outFile.puts "Performing immediate operation: #{sql}" 
		stmt = nil
		begin
			stmt = IBM_DB.exec(connection, sql)
		rescue
			outFile.puts "#{IBM_DB.stmt_error}"
			outFile.puts "#{IBM_DB.stmt_errormsg}"

		end
		unless stmt
			outFile.puts "Operation Failed"
		end
		return stmt
	end

	def setAutoComOff(connection, outFile)
		IBM_DB.autocommit(connection, IBM_DB::SQL_AUTOCOMMIT_OFF)
		
		# verify that in correct state
		ac = IBM_DB.autocommit(connection)
		if ac == 0
			outFile.puts "Autocommit set to off as intended"
		else
			outFile.puts "Cannot set autocommit to off.  FAIL"
		end
	end

	def rollbackTrans(connection, outFile)
		if IBM_DB.rollback(connection)
			outFile.puts "Rollback successful"
		else
			outFile.puts "Rollback FAILED"
		end
	end

	def commitTrans(connection, outFile)
		if IBM_DB.commit(connection)
			outFile.puts "Commit successful"
		else
			outFile.puts "Commit FAILED"
		end
	end

	def performPrepStmtBindings(connection, outFile, jLine, preparedStmtArray)
		currentBindings = jLine.bindingValueTuple
		outFile.puts "Values to bind to current prepared statement: #{currentBindings}"
		currentSqlOper = preparedStmtArray.getStmtIdSql(jLine.statementId)
		outFile.puts "Executing sql statement: #{currentSqlOper}"
		currentHdl = preparedStmtArray.getCurrentHdl(jLine.statementId)
		begin
			stmt = IBM_DB.execute(currentHdl, currentBindings)
		rescue
			outFile.puts "#{IBM_DB.stmt_error}"
			outFile.puts "#{IBM_DB.stmt_errormsg}"
		end
		unless stmt
			outFile.puts "Operation FAILED"
		end
	end

	def performPreparedStmt(connection, outFile, jLine, preparedStmtArray)
		currentSqlOper = getPreparedStmtSql(jLine.statementId, preparedStmtArray)
		outFile.puts "Executing sql statement: #{currentSqlOper}"
		currentHdl = getPreparedStmtHdl(jLine.statementId, preparedStmtArray)
		stmt = IBM_DB::execute currentHdl
		unless stmt
			outFile.puts "Operation FAILED"
		end
	end

	#  	ibm_db.bind_param documentation states this for value parameter
	# 	'A string specifying the name of the Ruby variable to bind to the parameter specified by parameter-number.''
	# 	The bind param operation is not performed until execute is performed.  If there is more than one parameter
	# 	they need different variable names and they must by in scope when execute ran.  For this automated testing
	# 	allowances need to be made to accommodate these requirements which would not be issue in normal use

	def prepareBindings(connection, outFile, jLine, preparedStmtArray)
		stmt = nil
		currentSqlOper = getPreparedStmtSql(jLine.statementId, preparedStmtArray)
		currentHdl = getPreparedStmtHdl(jLine.statementId, preparedStmtArray)
		index = 1
		jLine.bindings.each do |paramHash|
			$valueArray[index] = paramHash["value"]
			stmt = IBM_DB::bind_param currentHdl, paramHash["index"], "$valueArray[#{index}]", IBM_DB::SQL_PARAM_INPUT
			index += 1
		end
		unless stmt
			outFile.puts "Operation FAILED"
		end
	end

	def expectedEqualsActual(currentHdl, jLine, isOrdered, outFile)
		expectedResultsArray = jLine.expectedResults
		fetchNum = jLine.nfetch
		actualResultArray = getActualResultArray(currentHdl, fetchNum, outFile)
		outFile.puts "Expected Results: #{expectedResultsArray}"
		outFile.puts "Actual Results: #{actualResultArray}"
		if (actualResultArray == [] and expectedResultsArray == [])
        	return true
    	elsif isOrdered
        	return actualResultArray == expectedResultsArray
        else
        	found = false
        	actualResultArray.each do |entry|
        		expectedResultsArray.each do |hashRow|
        			if hashRow == entry
        				expectedResultsArray.delete(hashRow)
        				found = true
        				break
        			end
        		end
        		unless found
        			break
        		end
        	end
        	if expectedResultsArray.length > 0
        		found = false
        	end
        	if found
        		return true
        	else 
        		return false
        	end
        end
    end

    def getActualResultArray(stmtHdl, fetchNum, outFile)
    	resultArray = []
    	if fetchNum == nil
			resultHash = IBM_DB.fetch_assoc(stmtHdl)
			unless resultHash 
				return resultArray
			end
			while resultHash
				resultArray.push(resultHash)
				resultHash = IBM_DB.fetch_assoc(stmtHdl)
			end
			return resultArray
		else
			fetchNum.times do |i|
				resultHash = IBM_DB.fetch_assoc(stmtHdl)
				resultArray.push(resultHash)
			end
			return resultArray
		end
	end

	def getPreparedStmtSql(jLineStmtId, preparedStmtArray)
		preparedStmtArray.each do |preparedStmt|
			if preparedStmt.stmtId == jLineStmtId
				return preparedStmt.sqlStmt
			end
		end
	end

	def getPreparedStmtHdl(jLineStmtId, preparedStmtArray)
		preparedStmtArray.each do |preparedStmt|
			if preparedStmt.stmtId == jLineStmtId
				return preparedStmt.stmtHdl
			end
		end
	end

	def getPreparedStmtObj(jLineStmtId, preparedStmtArray)
		preparedStmtArray.each do |preparedStmt|
			if preparedStmt.stmtId == jLineStmtId
				return preparedStmt
			end
		end
	end

	def enterResultToJsonFile(fileName, result, outfile, comment)
		jsonHash = {}
		jsonHash[:language] = "ruby"
		jsonHash[:client] = "IBM DB2 ODBC DRIVER"
		jsonHash[:server] = "Informix 12.10"
		jsonHash[:test] = fileName
		jsonHash[:result] = result
		jsonHash[:detail] = comment
		outfile.puts(jsonHash.to_json)
	end

	def removePreparedStmtFromList(jLineStmtId, preparedStmtArray)
		preparedStmtArray.each do |preparedStmt|
			if preparedStmt.stmtId == jLineStmtId
				IBM_DB.free_stmt(preparedStmt.stmtHdl)
				preparedStmtArray.delete(preparedStmt)
				return preparedStmt.sqlStmt
			end
		end
	end

	def enterExtraDataToJsonFile(outfile)
		jsonHash = {}
	    jsonHash[:language] = "Ruby"
	    jsonHash[:client] = "IBM DB2 ODBC DRIVER"
	    jsonHash[:server] = "Informix 12.10"
	    jsonHash[:test] = "dataTypeTest_INT8.json"
	    jsonHash[:result] = false
	    jsonHash[:detail] = "int8 not compatible"  
	    outfile.puts(jsonHash.to_json)      
	    jsonHash = {}
	    jsonHash[:language] = "Ruby"
	    jsonHash[:client] = "IBM DB2 ODBC DRIVER"
	    jsonHash[:server] = "Informix 12.10"
	    jsonHash[:test] = "dataTypeTest_SERIAL8.json"
	    jsonHash[:result] = false
	    jsonHash[:detail] = "serial8 not compatible"       
	    outfile.puts(jsonHash.to_json)
	    jsonHash = {}
	    jsonHash[:language] = "Ruby"
	    jsonHash[:client] = "IBM DB2 ODBC DRIVER"
	    jsonHash[:server] = "Informix 12.10"
	    jsonHash[:test] = "ops.json"
	    jsonHash[:result] = true
	    jsonHash[:detail] = "format of test is different than others, but operations performed in other tests"       
	    outfile.puts(jsonHash.to_json)
	end

	def getComment(fileName)
		comment = nil
		if fileName.include?("BIGINT") || fileName.include?("BIGSERIAL") 
			comment = "Driver successfully places data in database, but extracts as string"
		elsif fileName.include?("BOOLEAN")
			comment = "Driver uses 0/1 for true/false, causing certain operations to respond incorrectly"
		elsif fileName.include?("DATETIME")
			comment = "Driver successfully places data in database, but extracts with fraction format"
		elsif fileName.include?("SMALLFLOAT")
			comment = "Data fetched from database converted improperly"
		else
			comment = nil
		end
		return comment
	end


end
